components/common/src/main/java/com/hotels/styx/server/HttpInterceptorContext.java

```diff
@@ -30,46 +32,23 @@
 
     private final boolean secure;
     private final InetSocketAddress clientAddress;
+    private final Executor executor;
 
     /**
      * Construct a new instance.
      *
      * @param secure true if the request was received via SSL
      * @param clientAddress address that request came from, or null if not-applicable
      */
-    public HttpInterceptorContext(boolean secure, InetSocketAddress clientAddress) {
+    public HttpInterceptorContext(boolean secure, InetSocketAddress clientAddress, Executor executor) {
         this.secure = secure;
         this.clientAddress = clientAddress; // intentionally nullable
-    }
-
-    /**
-     * Construct a new instance.
-     *
-     * @param clientAddress address that request came from, or null if not-applicable
-     */
-    public HttpInterceptorContext(InetSocketAddress clientAddress) {
-        this(false, clientAddress);
-    }
-
-    /**
-     * Construct a new instance.
-     *
-     * @param secure true if the request was received via SSL
-     */
-    public HttpInterceptorContext(boolean secure) {
-        this(secure, null);
-    }
-
-    /**
-     * Construct a new instance.
-     */
-    public HttpInterceptorContext() {
-        this(false, null);
+        this.executor = executor;
     }
 
     // TODO deprecate
     public static HttpInterceptor.Context create() {
-        return new HttpInterceptorContext();
+        return new HttpInterceptorContext(false, null, Executors.newSingleThreadExecutor());
```

Comment:
An `urlProbe` function in `HealthCheck.kt` calls `create` to create an "empty" context to be passed in to the health checked object. There appears to be other callers too.

Therefore, this change introduces an executor leak unless something is changed. Instead of changing the call sites, I would consider passing in a re-used executor from a static context. Like so:

```
    private static final globalExecutor = Executors.newSingleThreadExecutor();

    //    .. snip ..

    public static HttpInterceptor.Context create() {
        return new HttpInterceptorContext(false, null, globalExecutor);
    }

```

Reply:
OTOH, after discussing with @fantayeneh, this `create()` is should only really be used as convenience method for testing purposes. As such it can be removed from the class, and we can provide a support method in the test module itself. At minimum it should be marked with `@VisibleForTesting` annotation.

For this reason I'll change the call sites accordingly.

Reply:
I would get rid of this method altogether as we can just pass in a newSingleThreadExecutor without creating any thread leaks in our tests.

Reply:
Hi @dvlato I have now removed all calls to `HttpInterceptorContext.create()` from within the Styx project. I have still left the function there in case some external test code depends on in.

Reply:
Hi,  in this solution we still have to schedule new threads in several of the tests. Why don't we use the calling thread as the "Executor" when nothing has been passed?  I think it would make more sense and it's what we were doing previously, so tests should not be affected.

What I means is that we should be able to pass _Runnable::run_ as the third parameter to create() and in that case the Tests that were changed just to call the method with an executor could be rolled back (or at least changed to use the same 'calling thread executor')

Reply:
>  Why don't we use the calling thread as the "Executor" when nothing has been passed? I think it would make more sense and it's what we were doing previously, so tests should not be affected.

Because `HttpInterceptorContext.create()` is fundamentally publicly visible test method in a production class. As such it has to go. That's why I have marked it as deprecated and removed all calls to it from within styx code. 

> What I means is that we should be able to pass Runnable::run as the third parameter to create() and in that case the Tests that were changed just to call the method with an executor could be rolled back (or at least changed to use the same 'calling thread executor')

👍 Yes, this makes sense. This would stop the thread leaks from any dependant code, for now.